misc-scripts/xref_mapping/XrefParser/RefSeqGPFFParser.pm

```diff
@@ -17,340 +17,341 @@ limitations under the License.
+    }
   }
 
-  (%entrez)     = %{$self->get_acc_to_label("EntrezGene",$species_id, undef, $dbi)};
+  # populate entrez gene id => label hash
+  $self->{entrez} = $self->get_acc_to_label('EntrezGene', $species_id, undef, $dbi);
+
+  # get the species name, prepare species related data checks
+  my %species2name = $self->species_id2name($dbi);
+  $species_name //= shift @{$species2name{$species_id}};
+
+  $self->{name2species_id} = { map{ $_=>$species_id } @{$species2name{$species_id}} };
+  my %species2tax  = $self->species_id2taxonomy($dbi);
+  $self->{taxonomy2species_id} = { map{ $_=>$species_id } @{$species2tax{$species_id}} };
+
+  # process the source files
+  foreach my $file (@{$files}) {
 
-    my @xrefs;
-    foreach my $file (@files) {
-        my $xrefs =
-          $self->create_xrefs( $file, $species_id, $verbose, $dbi, $species_name );
+    # type from the file (peptide or dna)
+    my $type = $self->type_from_file($file);
 
-        if ( !defined( $xrefs ) ) {
-            return 1;    #error
+    # get the file handler
+    my $refseq_fh = $self->get_filehandle($file);
+
+    if ( !defined $refseq_fh ) {
+      warn "WARNING: Can't open RefSeqGPFF file $file\n";
+      return;
+    }
+
+    # this will hold the array of xrefs to bulk insert
+    my $xrefs;
+
+    do {
+      local $/ = "\/\/\n";
+      while ( my $item = $refseq_fh->getline() ) {
+        my $xref = $self->xref_from_record({
+          record => $item,
+          type   => $type
+        });
+
+        if (defined $xref) {
+          push @{$xrefs}, $xref;
         }
-        $self->upload_xref_object_graphs( $xrefs, $dbi )
+      }
+    };
+
+    $refseq_fh->close();
+
+    # no xrefs in this record...
+    if ( !defined( $xrefs ) ) {
+      next;
     }
 
-    if ( defined $release_file ) {
-        # Parse and set release info.
-        my $release_io = $self->get_filehandle($release_file);
-        local $/ = "\n*";
-        my $release = $release_io->getline();
-        $release_io->close();
-
-        $release =~ s/\s{2,}/ /g;
-        $release =~ s/.*(NCBI Reference Sequence.*) Distribution.*/$1/s;
-        # Put a comma after the release number to make it more readable.
-        $release =~ s/Release (\d+)/Release $1,/;
-
-        print "RefSeq release: '$release'\n" if($verbose);
-
-        $self->set_release( $source_id,              $release, $dbi );
-        $self->set_release( $peptide_source_id,      $release, $dbi );
-        $self->set_release( $mrna_source_id,         $release, $dbi );
-        $self->set_release( $ncrna_source_id,        $release, $dbi );
-        $self->set_release( $pred_mrna_source_id,    $release, $dbi );
-        $self->set_release( $pred_ncrna_source_id,   $release, $dbi );
-        $self->set_release( $pred_peptide_source_id, $release, $dbi );
+    # upload the xrefs
+    $self->upload_xref_object_graphs( $xrefs, $dbi );
+
+  }
+
+
+  # process the release file
```

Comment:
If the arguments specify a release file and this doesn't work - no release file, can't parse it, or something - it probably represents a critical failure due to something being very wrong, and I wonder if the parser shouldn't fail fast instead of at the end. Maybe the more robust sequence would be:
2) get release string
1) get and process xrefs
3) set release
instead of current 1,2,3.

Reply:
It's not a critical failure, it's not even a failure just a warning.
did set it to verbose only